### ISSUE TITLE
Lost Connection

### ISSUE THREAD

**Issue by MeepBeep** [jeffpiazza/derbynet#272]
I just started to run some tests and I'm getting a repeatable issue,...

Using the Champ Timer (4 lane) after heat 2 the timer within about a second blanks out the results and displays . . . . like it reset but then come heat 3 the results are not passed to the computer, if I manually add the times it will work for the next heat but generally after 2 or 3 (aka heat 7, 10, 13, 15) the issue repeats again and I have to manually enter those heats...

Any ideas?  I tried a different serial-USB adapter and same results...

If you need logs please provide me step by step and I will be happy to post them...  I tried getting logs but all they appear to show is a bunch of port busy...

+23:26:05.303S          INT openPort fails for COM6: Port busy
+23:26:15.318S          INT openPort fails for COM6: Port busy
+23:26:25.335S          INT openPort fails for COM6: Port busy
+23:26:35.342S          INT openPort fails for COM6: Port busy

**Comment by MeepBeep** [comment]
OK got the logging working, on this test race, heat 2 and heat 8 failed to record results and had to be manually entered...

[timer-4658873672365931556.log](https://github.com/jeffpiazza/derbynet/files/12912292/timer-4658873672365931556.log)

**Comment by jeffpiazza** [comment]
By default, the timer preparation for the next heat is delayed until at least 10s after the previous heat finishes.  (This is to allow the timer hardware to continue showing the times from the previous heat for a reasonable time.)

In actual racing, staging the next race easily takes longer than this, so it's not really a consideration, but in a testing situation, such as this, you may want to set -delay-reset-after-race to 0.

My guess is that you're sometimes running your simulated race within the 10s after the previous heat, and the timer hasn't been prepped.  E.g., heat 1 finishes at +00:04:48, then the reset commands to prep heat 2 happen at +00:04:58, then the heat gets canceled at +00:05:27 and heat 3 gets prepared immediately after.

If the delay is set to 0, you'll probably notice the timer display reset almost immediately after a heat completes.

**Comment by MeepBeep** [comment]
That makes a lot of sense and would explain the close but not exact 'pattern' of it repeating every so many test heats...

Time for more testing at a slower pace 🤣

**Comment by MeepBeep** [comment]
DerbyNet ran the race just fine as expected with the slower pace between heats, no issues until literally the last heat when the timer randomly disconnected for about 30 seconds and then popped back into action so we could wrap up the day 🤣

**Comment by jeffpiazza** [comment]
Normally, for hardware like the Champ that support a command set, the timer interface regularly polls the device, and all the kiosks will show an overlay if communication to the timer appears to be lost.  Is this what you observed, or merely a delay in presenting the heat results?

Any log files available?

**Comment by MeepBeep** [comment]
Yeah, all the kiosks went red lost timer overlay, and it was a Champ timer...  No logs...

When it happened I tested (jiggled) the connections at the timer and computer nothing changed, and forgive me if the next sequence of events is not perfect as I was under a little stress 🤣 but I believe even though all the kiosks went red overlay, in the Racedashboard the timer was still showing green and that was causing me confusion, then when I switched over to the ../timer.php window I believe it was yellow for a split second then went to green and that is when it reconnected and all the kiosks came back to life and we could finish...  And I fibbed about it being out for only 30 seconds, looking over the retrospect log is was probably a minute or two...

From the retrospect logging it clearly did lose connection...

Original heat completion
2023-10-22 14:35:03	
Timer heartbeat transmission overdue by 932ms
2023-10-22 14:35:04	
Timer Malfunction: No response from timer in 2214ms.
Cadette/Senior/Ambassador (R#4) heat 19	2023-10-22 14:39:27	
Original heat completion

**Comment by jeffpiazza** [comment]
Since you mentioned timer.php, I'm guessing you were running with the in-browser timer interface.

Browser type, and, if you have it, the version?  Running under which OS?

Could the laptop (or whatever it was) have gone to sleep at that moment?

**Comment by MeepBeep** [comment]
Yep, I was using the in-browser timer interface

Chrome 117.0.5938.152

Windows 10 Pro 22H2 64-bit (OS build 19045.3570) aka current with all updates as of Oct. 10

On a private dedicated LAN, no Internet, just the server and kiosks

It's a very clean and newer install of Windows, with basically just Chrome and Derbynet installed

No sleep mode, that was the the first thing I disabled.

As I said it ran fine all day until that small hiccup, we started racing at 9 am and that glitch was at 2:35 pm

Honestly no complaints, from loading the racers from an external database, to check in, to racer photos, to car photos and running the race it was a lot easier and smoother than I expected it would be...

It worked so well, that I will be using it for next years BSA race, as my race crew REALLY like the on-deck kiosk, that I had setup at the start of the track on a 10" tablet mounted to a tripod so there was no thought process in what cars to stage and what cars were racing and in what lane!  That alone sold it to them vs having to look at the 'backup' number stickers that I have seen traditionally leaned upon heavily to identify cars in other races...

**Comment by jeffpiazza** [comment]
It looks like the original complaint was attributable to the enforced (unless overridden) 10s delay after a heat's completion.

The last "lost connection" issue didn't leave any real clues, and appears not to have reappeared in subsequent use.

I'm going to close this issue, but feel free to reopen if you experience an unexplained lost connection again.
